### ISSUE TITLE
Tunnistaa kasvitaudit automaattisesti kuvista.

### ISSUE THREAD

**Issue by joonaris** [GispoCoding/eip-peltomappi#8]

**Comment by joonaris** [comment]
<html>
<body>
<html><head></head><body><p>Kyllä, <strong>kasvitauteja voi tunnistaa suoraan kuvasta Python-skriptin avulla</strong> hyödyntämällä <strong>koneoppimista (ML)</strong> ja erityisesti <strong>syväoppimismalleja (deep learning)</strong>. Tähän tarkoitukseen käytetään yleensä <strong>neuroverkkoja</strong>, jotka on koulutettu tunnistamaan kasveissa esiintyviä tauteja lehtien kuvista.</p>
<hr>
<h2>🧠 Yleinen lähestymistapa kasvitautien tunnistukseen Pythonilla:</h2>
<h3>1. 📸 Kuva-aineisto</h3>
<p>Tarvitset <strong>hyvin luokiteltua kuva-aineistoa</strong>: esim. lehtikuvia, joihin on liitetty tieto siitä, mikä tauti (jos mikään) niissä on.</p>
<p>Yksi tunnettu datasetti:</p>
<ul>
<li>
<p><strong>PlantVillage Dataset</strong> (Cornell University)</p>
<ul>
<li>
<p>Sisältää kymmeniä tuhansia kuvia eri kasveista ja taudeista.</p>
</li>
<li>
<p>Esim. tomaatti, peruna, maissi, jne.</p>
</li>
</ul>
</li>
</ul>
<h3>2. 🛠️ Teknologiat</h3>
<p>Python-kirjastoja:</p>
<ul>
<li>
<p><strong>TensorFlow</strong> tai <strong>PyTorch</strong>: neuroverkkopohjaiseen mallin koulutukseen</p>
</li>
<li>
<p><strong>OpenCV</strong>: kuvankäsittelyyn</p>
</li>
<li>
<p><strong>scikit-learn</strong>: yksinkertaisempiin ML-malleihin</p>
</li>
<li>
<p><strong>FastAI</strong>: nopeaan syväoppimisen käyttöönottoon (PyTorchin päällä)</p>
</li>
</ul>
<hr>
<h2>🧪 Esimerkki: yksinkertainen malli FastAI:lla</h2>
<pre><code class="language-python">from fastai.vision.all import *

# Ladataan kuvadata (oletetaan, että sinulla on kansiorakenne kuten 'tomaatti/healthy', 'tomaatti/blight', jne.)
path = Path("kasvikuva_datasetti")

# Luo data loaderin
dls = ImageDataLoaders.from_folder(
    path, 
    train=".", 
    valid_pct=0.2, 
    item_tfms=Resize(224), 
    batch_tfms=aug_transforms()
)

# Koulutetaan malli (ResNet on tehokas lähtökohta)
learn = vision_learner(dls, resnet34, metrics=accuracy)
learn.fine_tune(4)

# Ennustetaan yksittäisestä kuvasta
img = PILImage.create("testikuva.jpg")
pred_class, pred_idx, outputs = learn.predict(img)

print(f"Ennustettu tauti: {pred_class}")
</code></pre>
<hr>
<h2>🔍 Mallin tarkkuus</h2>
<ul>
<li>
<p>Oikein koulutettu malli voi yltää yli <strong>90 % tarkkuuteen</strong>, jos kuva-aineisto on laadukasta.</p>
</li>
<li>
<p>Hyvät valaistusolosuhteet ja selkeät kuvat parantavat ennustustulosta.</p>
</li>
</ul>
<hr>
<h2>🧰 Vaihtoehtoja ilman omaa mallia</h2>
<p>Jos et halua kouluttaa itse:</p>
<ul>
<li>
<p>On olemassa <strong>valmiita malleja ja API-palveluita</strong>, esim. PlantDoc-malli, Plantix, AgriNet.</p>
</li>
<li>
<p>Jotkin avoimen lähdekoodin projektit tarjoavat esikoulutettuja malleja, joita voi käyttää suoraan Pythonissa.</p>
</li>
</ul>
<hr>
<h2>⚠️ Huomioita</h2>
<ul>
<li>
<p>Mallit voivat <strong>sekoittaa mekaanisia vaurioita, ravinnepuutoksia ja tauteja</strong>.</p>
</li>
<li>
<p>Jos haluat käyttää kuvia <strong>droonista tai satelliitista</strong>, tarvitaan erilaista mallia ja usein <strong>hyperspektridata</strong>.</p>
</li>
</ul>
<hr>
<h2>✅ Yhteenveto</h2>

Vaihe | Kuvaus
-- | --
1. | Hanki laadukas, luokiteltu kuva-aineisto (esim. PlantVillage)
2. | Käytä syväoppimismallia (FastAI, TensorFlow, PyTorch)
3. | Kouluta malli tai käytä valmista
4. | Syötä uusia kuvia ja tulkitse tulokset


<hr>
<p>Jos kerrot tarkemmin, <strong>minkä kasvin tauteja</strong> haluat tunnistaa ja <strong>millaisia kuvia</strong> on käytettävissä (puhelinkuvat? dronella? resoluutio?), voin auttaa vielä täsmällisemmin — esim. antaa valmiin mallin tai tehdä sinulle pohjan projektin aloittamiseen.</p></body></html>
</body>
</html>

**Comment by joonaris** [comment]
<html>
<body>
<html><head></head><body><p>QGISiin voidaan <strong>yhdistää kasvitautien tunnistusmalli Pythonilla</strong> eri tavoin — helpoimmasta täysin automatisoituun integraatioon. Tässä käyn läpi <strong>3 toteutustapaa</strong>, joilla saat <strong>kasvitautien tunnistuksen toimimaan QGISin sisällä tai sen kanssa</strong>.</p>
<hr>
<h2>🔁 Tapa 1: <strong>Python-skripti + QGIS Processing -työkalu</strong></h2>
<p>Tämä on suositeltava lähestymistapa: teet Pythonilla skriptin, joka toimii <strong>Processing-työkaluna QGISissä</strong> ja käsittelee kuvan -&gt; antaa tuloksen -&gt; kirjoittaa se kartalle (esim. attribuuttikenttään tai popup-tekstiksi).</p>
<h3>🛠️ Vaiheet:</h3>
<ol>
<li>
<p>Kirjoita kasvitautien tunnistusskripti (FastAI, TensorFlow jne.).</p>
</li>
<li>
<p>Tee siitä <strong>QGISin Processing-skripti</strong> (käytännössä <code inline="">.py</code>-tiedosto).</p>
</li>
<li>
<p>QGISissä: <code inline="">Processing &gt; Toolbox &gt; Create new script</code>.</p>
</li>
</ol>
<h3>📄 Esimerkki (yksinkertaistettu):</h3>
<pre><code class="language-python">##Kasvitauti: kasvitauti_analyysi
##Kuva=tiedosto

from fastai.vision.all import *

def run(Kuva):
    # Lataa valmis malli (esim. export.pkl FastAI:sta)
    learn = load_learner("malli/export.pkl")
    pred_class, _, _ = learn.predict(Kuva)
    
    print(f"Kasvitautiluokitus: {pred_class}")
</code></pre>
<p>Kun tallennat tämän <code inline="">kasvitauti_analyysi.py</code>-nimellä <code inline="">QGIS/scripts</code>-kansioon, se näkyy QGISissä omana työkalunaan. Voit käyttää sitä minkä tahansa rasterikuvan kanssa.</p>
<hr>
<h2>🔄 Tapa 2: <strong>Attribuuttitoiminto tai lomakelogiikka (käynnistää Pythonin taustalla)</strong></h2>
<p>Voit käyttää <strong>QGISin mukautettuja lomakkeita</strong> (Qt Designer + Python), joissa lisätään esim. "Analysoi kuva" -nappi, joka:</p>
<ul>
<li>
<p>Käynnistää kasvitautien analyysin</p>
</li>
<li>
<p>Näyttää tuloksen</p>
</li>
<li>
<p>Tallentaa sen attribuuttikenttään tai layerin popupiin</p>
</li>
</ul>
<h3>Esimerkki:</h3>
<ul>
<li>
<p>Kuvakohde valitaan kartalta</p>
</li>
<li>
<p>Lomakkeessa painetaan "Analysoi kuva"</p>
</li>
<li>
<p>Tulos: <code inline="">Perunarutto</code> → tallennetaan <code inline="">tauti</code>-kenttään</p>
</li>
</ul>
<p>Tähän tarvitaan:</p>
<ul>
<li>
<p>Python-logiikkaa lomakkeessa (<code inline="">form.py</code>)</p>
</li>
<li>
<p>Qt Designerilla tehty lomake</p>
</li>
</ul>
<hr>
<h2>🧠 Tapa 3: <strong>Automaatio komentoriviltä + tulosten tuonti kartalle</strong></h2>
<p>Helpoin alku:</p>
<ol>
<li>
<p>Kirjoita erillinen Python-skripti, joka:</p>
<ul>
<li>
<p>Ottaa kuvatiedoston tai -kansion</p>
</li>
<li>
<p>Analysoi sen</p>
</li>
<li>
<p>Kirjoittaa tulokset CSV/GeoJSON/GPKG -tiedostoon</p>
</li>
</ul>
</li>
<li>
<p>QGISissä:</p>
<ul>
<li>
<p>Lataa tämä tulosaineisto kartalle</p>
</li>
<li>
<p>Liitä vaikka ilmakuvien tai paikkatietokerrosten päälle</p>
</li>
</ul>
</li>
</ol>
<p>Esimerkiksi: <code inline="">analyysi.py</code> tulostaa:</p>
<pre><code class="language-csv">paikka_x, paikka_y, tauti
273533, 6789231, Blight
273700, 6789100, Healthy
</code></pre>
<p>Tuo CSV QGISiin ja visualisoi pisteet + tautiluokat eri väreillä.</p>
<hr>
<h2>🧩 Bonus: mahdollinen live-analyysi pluginiin</h2>
<p>Jos haluat <strong>täydellisen integraation</strong>:</p>
<ul>
<li>
<p>Luo oma <strong>QGIS Python -plugin</strong> (käytetään Qt Designeria ja PyQt5/PyQt6:ta)</p>
</li>
<li>
<p>Plugin voi käyttää mallia taustalla</p>
</li>
<li>
<p>Näyttää tulokset suoraan kartalla tai popupissa</p>
</li>
</ul>
<p>Tämä on monimutkaisempi, mutta onnistuu hyvin, jos haluat ammattilaatuisen työkalun.</p>
<hr>
<h2>🧪 Käytännön esimerkki: FastAI + QGIS Processing -työkalu</h2>
<h3>Hakemistorakenne:</h3>
<pre><code>kasvitauti_analyysi/
│
├── kasvitauti_analyysi.py       ← QGIS Processing -työkalu
├── malli/
│   └── export.pkl               ← FastAI-malli
</code></pre>
<p>QGISin kautta valitset kuvan → saat tautiluokan → voit sijoittaa sen kartalle (esim. alueeseen liittyvä ominaisuus).</p>
<hr>
<h2>✅ Yhteenveto: miten yhdistää kasvitautimalli QGISiin</h2>

Tapa | Kuvaus | Soveltuu
-- | -- | --
🧩 QGIS Processing -skripti | Luo oma työkalu, käyttää mallia taustalla | Paras yhdistelmä yksinkertaisuutta ja tehokkuutta
🧠 Lomakkeen Python-logiikka | Lomakeanalyysi, esim. napista "Analysoi" | Jos haluat hienon käyttöliittymän
🔁 Ulkoinen analyysi + tulosten tuonti | CSV / GeoJSON -tulokset tuodaan QGISiin | Nopein aloitus, vähemmän integroitu
🧱 QGIS Plugin | Täysi lisäosa, analysoi ja piirtää suoraan | Ammattilaiskäyttöön, enemmän työtä


<hr>
<p>Haluatko että teen sinulle <strong>valmiin QGIS Processing -työkalun esimerkin</strong> (ja testikuvan + mallin käyttämiseksi), vai mietitkö jotain tiettyä kasvia tai dataa?</p></body></html>
</body>
</html>
